// contracts-db/index.js
const SDK = require('dat-sdk')
const { Hypercore, Hyperdrive, resolveName, deleteStorage, destroy } = SDK()

module.exports = contractsDB

function contractsDB (daturl) {
  const contracts = []
  let waitingQuery = null
  let ready = false

  // ADD MOLOCH DAO - for the demo @TODO remove after the demo
  const moloch = require('./moloch-demo.sol')
  contracts.push({
    source: moloch,
    title: 'Moloch.sol',
    hash: '0x1234567345678903456'
  })

  var counter = 0
  const archive = Hyperdrive(daturl)
  window.archive = archive
  return { getAll, get }

  function get (query, done) {
    if (!ready) {
      waitingQuery = [query, done]
      return
    }
    const match = []
    const formattedSources = [...contracts]
    for (var i = 0; i < contracts.length; i++) {
      const temp = formattedSources[i].source.replace(/\n. |\r/g, "")
      const target = formattedSources[i]
      if (temp.includes(query)) {
        match.push({
          source: target.source,
          title: target.title,
          hash: target.hash
        })
      }
    }
    done(match)
  }
    // retrieve all sources
  function getAll (done) {
    console.log('Retreiving data from the P2P database')
    archive.ready(() => {
      archive.readdir('.', (err, addresses) => {
        if (err) return done(err)
        if (addresses) {
          counter = addresses.length
          for (var i=0; i<addresses.length; i++) {
            getData (addresses, i, done)
          }
        }
      })
    })
  }
  // get data
  function getData (addresses, i, done) {
    const filepath = `${addresses[i]}`
    archive.readFile(filepath, 'utf8', (err, result) => {
      if (err) return done(err) // console.log(err)
      data = JSON.parse(result)
      contracts.push({
        source: data.sourceCode,
        title: data.contractName,
        hash: data.address })
      console.log(`Contracts retreived: ${contracts.length}`)
      if (counter === contracts.length) {
        ready = true
        if (waitingQuery) get(waitingQuery[0], waitingQuery[1])
        done(null, contracts)
      }
    })
  }
}
